### PR TITLE
Add search history provider

### PR DESCRIPTION
**Pull request by markwu**
I just add a search history support, it copied from command_history.vim.

### CODE REVIEW

File: autoload/clap/provider/search_history.vim [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+let s:save_cpo = &cpoptions
+set cpoptions&vim
+
+" Derived from fzf.vim
+function! s:search_history() abort
+  let max  = histnr('/')
+  let fmt  = ' %'.len(string(max)).'d '
+  let list = filter(map(range(1, max), 'histget(":", - v:val)'), '!empty(v:val)')
+  return list
```
Comment (liuchengxu) [inline autoload/clap/provider/search_history.vim]: Try refactoring the duplicated code.

Comment (markwu) [reply autoload/clap/provider/search_history.vim]: Okay!

File: autoload/clap/provider/search_history.vim [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+function! s:search_history_sink(selected) abort
+  let item = matchstr(a:selected, '\d\+\s\+\zs\(.*\)')
+  let @/ = item
+  execute 'normal n'
+endfunction
+
+let s:search_history = {}
+let s:search_history.sink = function('s:search_history_sink')
+let s:search_history.source = function('s:search_history_source')
+let s:search_history.syntax = 'clap_search_history'
```
Comment (liuchengxu) [inline autoload/clap/provider/search_history.vim]: If you use the same syntax file with `clap_search_history`,no need to copy `syntax/clap_command_history`, just set the `syntax` to `clap_command_history`, even though the semantic is not quite right, we don't care as we merely want the highlight.

Comment (markwu) [reply autoload/clap/provider/search_history.vim]: okay.

File: autoload/clap/provider/search_history.vim (hunk 1 above)
Comment (markwu) [reply autoload/clap/provider/search_history.vim]: Done.

File: autoload/clap/provider/search_history.vim (hunk 2 above)
Comment (markwu) [reply autoload/clap/provider/search_history.vim]: Done!

File: autoload/clap/provider/common_history.vim [hunk 3]
```diff
@@ -0,0 +1,33 @@
+" Author: liuchengxu <xuliuchengxlc@gmail.com>
+" Description: Common utilities for command/search history.
```
Comment (liuchengxu) [inline autoload/clap/provider/common_history.vim]: Do not place the utils code in the provider directory, which will be considered as a provider incorrectly by `:Clap providers`. You can put these code in `autoload/clap/util.vim`.

Comment (markwu) [reply autoload/clap/provider/common_history.vim]: Ah ... Okay!

File: autoload/clap/common_history.vim [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  return list
+endfunction
+
+function! clap#common_history#source(name) abort
+  let hist_list = s:get_history_list(a:name)
+  let hist_len = len(hist_list)
+  return map(hist_list, 'printf("%4d", hist_len - v:key)."  ".v:val')
+endfunction
+
+function! clap#common_history#sink(name, selected) abort
```
Comment (liuchengxu) [inline autoload/clap/common_history.vim]: Is `type` a better name?
```suggestion
function! clap#common_history#sink(type, selected) abort
```

Comment (markwu) [reply autoload/clap/common_history.vim]: My original variable name is `type`, but it located in  `hist-names` section in vim help. That's why I change it to `name`.

I am okay to change it back to `type`.

Comment (liuchengxu) [reply autoload/clap/common_history.vim]: >but it located in hist-names section in vim help.

I don't follow.

File: autoload/clap/provider/common_history.vim (hunk 3 above)
Comment (markwu) [reply autoload/clap/provider/common_history.vim]: Done!

File: autoload/clap/common_history.vim (hunk 4 above)
Comment (markwu) [reply autoload/clap/common_history.vim]: Done!

Comment (markwu) [reply autoload/clap/common_history.vim]: I search the history names or types in vim help, it calls them `hist-names`.  Try `:help hist-names` you will see.

```vim
		one of:					*hist-names*
			"cmd"	 or ":"	  command line history
			"search" or "/"   search pattern history
			"expr"	 or "="   typed expression history
			"input"  or "@"	  input line history
			"debug"  or ">"   debug command history
			empty		  the current or last used history
```

File: autoload/clap/common_history.vim [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  let hist_len = len(hist_list)
+  return map(hist_list, 'printf("%4d", hist_len - v:key)."  ".v:val')
+endfunction
+
+function! clap#common_history#sink(type, selected) abort
+  let item = matchstr(a:selected, '\d\+\s\+\zs\(.*\)')
+  if a:type ==# ':'
+    call histadd(':', item)
+    call execute(item, '')
+  elseif a:type ==# '/'
```
Comment (liuchengxu) [inline autoload/clap/common_history.vim]: Have you tested yourself?  `/` sink does not always work as expected, the editor can be freezed and I have to press `<C-C>` to exit.

Comment (markwu) [reply autoload/clap/common_history.vim]: Yes, I tested. I never meet this issue before. Can you reproduce it?

Comment (liuchengxu) [reply autoload/clap/common_history.vim]: My issue is due to `n` is mapped in my config.

```diff
diff --git a/autoload/clap/common_history.vim b/autoload/clap/common_history.vim
index 444eb11..0efbc8b 100644
--- a/autoload/clap/common_history.vim
+++ b/autoload/clap/common_history.vim
@@ -25,7 +25,7 @@ function! clap#common_history#sink(type, selected) abort
     call execute(item, '')
   elseif a:type ==# '/'
     let @/ = item
-    execute 'normal n'
+    normal! n
   endif
 endfunction
```

Using `normal!` fixes my issue, a plugin should usually should use `normal!` for avoiding such issues.

Comment (markwu) [reply autoload/clap/common_history.vim]: Ah, okay!

Comment (markwu) [reply autoload/clap/common_history.vim]: Done!

**Comment by liuchengxu** [comment]
Merged, thanks!

**Pull request merged**